### ISSUE TITLE
[FEATURE] Supports sorting of data in the Elasticsearch vector database

### ISSUE THREAD

**Issue by t19980620** [langchain4j/langchain4j#1640]
When Elasticsearch is used as a vector database to retrieve data, it is not possible to sort the Elasticsearch data. For example, sorting by the age field

**Comment by langchain4j** [comment]
cc @dadoonet (do you mind if I'll ping you on Elasticsearch issues?)

**Comment by dadoonet** [comment]
Sure! Please do or assign me 😊

**Comment by dadoonet** [comment]
About this feature request: I guess this is something which needs some update in the `EmbeddingSearchRequest` class. We would need to add some sort orders on a list of fields. In Elasticsearch we support things like:

```java
.sort(so -> so
        .geoDistance(gd -> gd
            .field("location")
            .location(
                    new GeoLocation.Builder()
                            .latlon(ll -> ll.lat(49.0404).lon(2.0174))
                            .build()
            )
        .order(SortOrder.Asc)
        .unit(DistanceUnit.Kilometers)
)
```

How should we design some generic sorting in LangChain4j? Any thoughts?

**Comment by langchain4j** [comment]
@dadoonet good question. Perharps sorting can be configured in the builder of `ElasticsearchEmbeddingStore` for the moment. And, if/when it will be required in other stores, we can think about adding it to the common API, WDYT?

**Comment by dadoonet** [comment]
Yeah. We can do that probably. Although that will be very "static" if the user wants to change the order at request time.

Think of an e-commerce website where the user want dynamically order by price, note or date. We probably don't want to recreate an instance of the embedding store for each call/user.

**Comment by langchain4j** [comment]
@dadoonet if dynamism is required, `ElasticsearchEmbeddingStore.search(EmbeddingSearchRequest, Sorting)` or similar can be added.

@t19980620 could you please share a bit more info on your use case? Thanks!

**Comment by dadoonet** [comment]
I'd keep `ElasticsearchEmbeddingStore.search(EmbeddingSearchRequest)` but change the `EmbeddingSearchRequest` object to add a `sort` field and its builders.

```java
EmbeddingSearchRequest.builder()
  .queryEmbedding(queryEmbedding)
  .sort(Sort.builder()
      .field("foo").order(Sort.ASC)
      .build())
  .build());
```

With that, we don't have to change all the existing code. embedding stores will "just" have to modify their implementation of `search(EmbeddingSearchRequest)` when they want to support sorting.

**Comment by t19980620** [comment]
> @dadoonet if dynamism is required, `ElasticsearchEmbeddingStore.search(EmbeddingSearchRequest, Sorting)` or similar can be added.
> 
> @t19980620 could you please share a bit more info on your use case? Thanks!

The following is an actual business scenario for searching with the native elasticseartch-java client, which I hope langchain4hj can also support

```java
elasticsearchClient.search(s -> s.index("material_video" + esIndexSuffix)
                                .knn(KnnQuery.of(m -> {
                                    BoolQuery.Builder builder = QueryBuilders.bool()
                                            .must(EsUtils.matchPhrase("biz", CategoryBizEnum.GOODS.name()));
                                    return m.field("video_vector")
                                            .queryVector(JsonUtils.toList(screenDescriptionVector, Float.class))
                                            .k(5)
                                            .numCandidates(100)
                                            .filter(builder.build()._toQuery());
                                }))
                                .sort(sort -> sort.field(fs -> fs.field("use_num").order(SortOrder.Asc)))
                                .from(0)
                                .size(1),
                        MaterialVideoEsDO.class);
```
Especially the `k()` and `numCandidates()` parameters, I hope they can be set externally

**Comment by dadoonet** [comment]
> Especially the `k()` and `numCandidates()` parameters, I hope they can be set externally

So this is another ask than the sort feature.
Could you open another issue for this?

I think it's much easier to implement and this should be part IMHO of the configuration. So much easier.

**Comment by langchain4j** [comment]
@dadoonet

> I'd keep ElasticsearchEmbeddingStore.search(EmbeddingSearchRequest) but change the EmbeddingSearchRequest object to add a sort field and its builders.

I am not yet convinced that sorting is the feature that should be included in the `EmbeddingSearchRequest`. In the LLM/RAG context, sorting is usually done by similarity, so this seems like an edge case to me. If/when it will be proven that this is a feature that multiple users require, we can introduce a new field in the `EmbeddingSearchRequest` and support if for more stores. But for now I would keep it only in `ElasticsearchEmbeddingStore`.

> With that, we don't have to change all the existing code. embedding stores will "just" have to modify their implementation of search(EmbeddingSearchRequest) when they want to support sorting.

There is no need to change anything now, only the `ElasticsearchEmbeddingStore` implementation. If/when sorting will be added to the `EmbeddingSearchRequest`, changes in `ElasticsearchEmbeddingStore` will be minimal.

**Comment by dadoonet** [comment]
I created #1649 to support the `num_candidates` [feature](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-knn-query.html#knn-query-top-level-parameters).
Ideally, we could bring this in the 0.34 release. ;)
